File: app/src/main/java/com/norvera/confession/ui/main/MainViewModel.java
package com.norvera.confession.ui.main;

import com.norvera.confession.data.AppRepository;
import com.norvera.confession.data.models.CommandmentEntry;
import com.norvera.confession.data.models.ExaminationActiveEntry;
import com.norvera.confession.data.models.ExaminationEntry;
import com.norvera.confession.data.models.GuideEntry;
import com.norvera.confession.data.models.InspirationEntry;
import com.norvera.confession.data.models.PrayersEntry;
import com.norvera.confession.data.models.User;

import java.util.List;

import androidx.databinding.ObservableField;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.sqlite.db.SimpleSQLiteQuery;
import timber.log.Timber;

public class MainViewModel extends ViewModel {
    private static final String TAG = MainViewModel.class.getSimpleName();

    private final AppRepository appRepository;



    public ObservableField<List<ExaminationEntry>> examinationEntries = new ObservableField<>();

    MainViewModel(AppRepository repository) {
        Timber.d("Actively retrieving the movies from the DataBase");
        appRepository = repository;
    }

    public LiveData<List<CommandmentEntry>> allCommandmentsForLay() {
        return appRepository.loadCommandmentsForLayPerson();
    }

    public LiveData<List<CommandmentEntry>> allCommandmentsForReligious() {
        return appRepository.loadCommandmentsForReligious();
    }

    public LiveData<List<ExaminationEntry>> allExaminationsForCommandmentAndUser(SimpleSQLiteQuery query) {
        return appRepository.loadAllExaminationsForCommandmentAndUser(query);
    }

    public void updateCountForEntry(ExaminationEntry examinationEntry) {
        appRepository.updateCountForEntry(examinationEntry);
    }

    public void updateCountForEntryId(long examinationId) {
        //todo implement and test out
    }

    public LiveData<List<GuideEntry>> allGuidesForId(int guideId) {
        return appRepository.allGuidesForId(guideId);
    }

    public LiveData<List<PrayersEntry>> loadAllPrayers() {
        return appRepository.loadAllPrayers();
    }

    public LiveData<List<ExaminationActiveEntry>> allExaminationsWithCount() {
        return appRepository.loadAllExaminationsWithCount();
    }

    public LiveData<InspirationEntry> getInspirationForId(int id) {
        return appRepository.getInspirationForId(id);
    }

    public void decrementCount(ExaminationEntry examinationEntry) {
        appRepository.decrementCountForEntry(examinationEntry);
    }
}

File: app/src/main/java/com/norvera/confession/viewmodels/MainViewModel.kt
package com.norvera.confession.viewmodels

import androidx.databinding.ObservableField
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.sqlite.db.SimpleSQLiteQuery
import com.beautycoder.pflockscreen.security.PFResult
import com.beautycoder.pflockscreen.security.PFSecurityManager
import com.norvera.confession.data.AppRepository
import com.norvera.confession.data.models.*
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import timber.log.Timber

class MainViewModel internal constructor(private val appRepository: AppRepository) : ViewModel() {


    var examinationEntries = ObservableField<List<ExaminationEntry>>()
    private val liveData = MutableLiveData<PFResult<Boolean>>()//PFLiveData<PFResult<Boolean>>()

    init {
        Timber.d("Actively retrieving the movies from the DataBase")
    }
    /**
     * This is the job for all coroutines started by this ViewModel.
     *
     * Cancelling this job will cancel all coroutines started by this ViewModel.
     */
    private val viewModelJob = Job()

    /**
     * This is the scope for all coroutines launched by [PlantDetailViewModel].
     *
     * Since we pass [viewModelJob], you can cancel all coroutines launched by [viewModelScope] by calling
     * viewModelJob.cancel().  This is called in [onCleared].
     */
    //private val viewModelScope = CoroutineScope(Dispatchers.Main + viewModelJob)

    /**
     * Cancel all coroutines when the ViewModel is cleared.
     */

    fun allCommandmentsForLay(): LiveData<List<CommandmentEntry>> {
        return appRepository.loadCommandmentsForLayPerson()
    }

    fun allCommandmentsForReligious(): LiveData<List<CommandmentEntry>> {
        return appRepository.loadCommandmentsForReligious()
    }

    fun allExaminationsForCommandmentAndUser(query: SimpleSQLiteQuery): LiveData<List<ExaminationEntry>> {
        return appRepository.loadAllExaminationsForCommandmentAndUser(query)
    }

    fun updateCountForEntry(examinationEntry: ExaminationEntry) {
        viewModelScope.launch {
            appRepository.updateCountForEntry(examinationEntry)
        }
    }

    fun allGuidesForId(guideId: Int): LiveData<List<GuideEntry>> {
        return appRepository.allGuidesForId(guideId)
    }

    fun loadAllPrayers(): LiveData<List<PrayersEntry>> {
        return appRepository.loadAllPrayers()
    }

    fun allExaminationsWithCount(): LiveData<List<ExaminationActiveEntry>> {
        return appRepository.loadAllExaminationsWithCount()
    }

    fun getInspirationForId(id: Int): LiveData<InspirationEntry> {
        return appRepository.getInspirationForId(id)
    }

    fun decrementCount(examinationEntry: ExaminationEntry) {
        viewModelScope.launch {
            appRepository.decrementCountForEntry(examinationEntry)
        }
    }

    fun isPinCodeEncryptionKeyExist(): LiveData<PFResult<Boolean>> {
        PFSecurityManager.getInstance().pinCodeHelper.isPinCodeEncryptionKeyExist { liveData.value = it }
        return liveData
    }


    companion object {
        private val TAG = MainViewModel::class.java.simpleName
    }
}
